cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.ql

```diff
@@ -12,29 +12,60 @@
 
 import cpp
 import semmle.code.cpp.commons.Exclusions
-import semmle.code.cpp.valuenumbering.GlobalValueNumbering
 import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+import semmle.code.cpp.rangeanalysis.RangeAnalysisUtils
 import semmle.code.cpp.controlflow.Guards
+import semmle.code.cpp.dataflow.DataFlow
 
-/** Holds if `sub` is guarded by a condition which ensures that `left >= right`. */
+/**
+ *  Holds if `sub` is guarded by a condition which ensures that
+ * `left >= right`.
+ */
 pragma[noinline]
 predicate isGuarded(SubExpr sub, Expr left, Expr right) {
-  exists(GuardCondition guard |
-    guard.controls(sub.getBasicBlock(), true) and
-    guard.ensuresLt(left, right, 0, sub.getBasicBlock(), false)
+  exists(GuardCondition guard, int k |
+    guard.controls(sub.getBasicBlock(), _) and
+    guard.ensuresLt(left, right, k, sub.getBasicBlock(), false) and
+    k >= 0
   )
 }
 
-/** Holds if `sub` will never be negative. */
-predicate nonNegative(SubExpr sub) {
-  not exprMightOverflowNegatively(sub.getFullyConverted())
+/**
+ * Holds if `n` is known or suspected to be less than or equal to
+ * `sub.getLeftOperand()`.
+ */
+predicate exprIsSubLeftOrLess(SubExpr sub, DataFlow::Node n) {
+  n = DataFlow::exprNode(sub.getLeftOperand())
+  or
+  exists(DataFlow::Node other |
+    // dataflow
+    exprIsSubLeftOrLess(sub, other) and
+    (
+      DataFlow::localFlowStep(n, other) or
```

Comment:
This part of the disjunction has me somewhat confused. It seems to specify that
```
if other <= sub.getLeftOperand() and n flows to other then n <= sub.getLeftOperand().
```

I thought I could use this to construct an false-negative example using something like this:
```cpp
void test(unsigned a, unsigned b) {
  unsigned other = a; // other <= a
  other = b; // b flows to other
  return a - b > 0; // now we conclude b <= a
}
```

However, we still seem to flag this. So I suppose it's not as bad as I initially thought.

Reply:
I think dataflow in either direction suggests that `n` and `other` are equal (it might be more natural to use GVN for this of course; in fact I believe I did use GVN in this query at some point).

Reply:
Yeah, I think dataflow in either direction suggests that there is some control-flow path that would make `n` and `other` equal.

In an ideal world, I would have expected GVN to also give this guarantee, but with the added bonus that `n` and `other` are equal in _every_ control-flow path. But seeing as this predicate is used to rule out alerts, I guess it's better to lean on the side of too many results.